adagrams/game.py

```diff
@@ -1,11 +1,139 @@
 def draw_letters():
-    pass
+    """
+        output: list of 10 strings
+        no parameters
+        letters should be randomly drawn from a pool of letters
+        letters should reflect distribution from dictionary (letters_dict)
+        drawn letters cannot return more than the allotted amount
+        once something is drawn, allotment will be subtracted by 1
+        if the value is 0, it is no longer accessible
+        """
+    hand_list = []
+    hand_dict = letters_dict.copy()
+    while len(hand_list) < 10:
+        random_letter = random.choice(string.ascii_uppercase)
+        for letter in hand_dict:
+            if letter == random_letter:
+                if hand_dict[letter] >= 1:
+                    hand_list.append(letter)
+                    hand_dict[letter] -= 1  
+    return hand_list
 
 def uses_available_letters(word, letter_bank):
-    pass
+    """
+    input: word (string), letter_bank: (list of drawn letters)
+    output: True or False
+    True if every letter in word is available in letter_bank
+    False if not above, or if there is a letter in word that is not present in the letter_bank,
+    or if word is longer than letter_bank
+    """
+    available_letters = letter_bank.copy()
+    if len(word) > len(letter_bank):
+        return False
+
+    extra_letters = set(word.upper()) - set(letter_bank)
+    if extra_letters:
+        return False
+
+    for letter in word.upper():
+        if not letter in available_letters:
+            return False
+        else:
+            available_letters.remove(letter)
+    return True
 
 def score_word(word):
-    pass
+    """
+    input: word (string)
+    output: integer (number of points scored)
+    function will add points assciated w/each letter
+    8 bonus points will be awarded if length of word >= 7
+    """
+    score = 0
+    
+    if len(word) >= 7:
+        score += 8
+    for letter in word.upper():
+        for key, value in score_dict.items():
+            if letter in value:
+                score += key
+
+    return score            
 
 def get_highest_word_score(word_list):
-    pass
\ No newline at end of file
+    """
+    input: list of strings
+    output: returns tuple winning_word (string), score (int)
+    in case of tie:
+    word with fewest letters will win
+    if multiple words have same length, winner is first one in list
+    exception: in case of tie, if one of the words has length of 10,
+    that word wins
+    """
+    score_list = []
+    tuple_list = []
+    for word in word_list:
+        word_tuple = word, score_word(word)
+        tuple_list.append(word_tuple)
+        score_list.append(score_word(word))
+
+    # finding max score
+    max_score = max(score_list)
+    max_index_list = [i for i, score in enumerate(score_list) \
+        if score == max_score]
+
+    if len(max_index_list) == 1:
+        return tuple_list[max_index_list[0]]
+    else:
+        small_word_length = 10
+        for word, score in tuple_list:
```

Comment:
Everything here works, but there is some room for optimization.  This loop iterates over every word/score tuple, but it only needs to iterate over the word/score tuple where score == max_score.  I suggest thinking about how to use the list comprehension on line 123 to construct a list of the tuples with the max score, rather than the indexes of those scores, and then use that list in the conditional in line 126 and in this loop.